pandas/_libs/tslibs/timezones.pyx

```diff
@@ -275,3 +275,84 @@ cdef object get_dst_info(object tz):
         dst_cache[cache_key] = (trans, deltas, typ)
 
     return dst_cache[cache_key]
+
+
+def _infer_tzinfo(start, end):
+    def _infer(a, b):
+        tz = a.tzinfo
+        if b and b.tzinfo:
+            if not (get_timezone(tz) == get_timezone(b.tzinfo)):
+                raise AssertionError('Inputs must both have the same timezone,'
+                                     ' {timezone1} != {timezone2}'
+                                     .format(timezone1=tz, timezone2=b.tzinfo))
+        return tz
+
+    tz = None
+    if start is not None:
+        tz = _infer(start, end)
+    elif end is not None:
+        tz = _infer(end, start)
+    return tz
+
+
+cdef ndarray[int64_t] _infer_dst(ndarray[int64_t] vals,
+                                 ndarray[int64_t] result_a,
+                                 ndarray[int64_t] result_b):
+    cdef:
+        Py_ssize_t n = len(vals)
+        ndarray[int64_t] dst_hours
```

Comment:
make sure u r typing s the original
there are lots of issues here

Reply:
The typings should be identical.  There are a couple of variables that do not have type declarations; those can be added.